src/Fleck/WebSocketServer.cs

```diff
@@ -93,34 +95,54 @@ public void Start(Action<IWebSocketConnection> config)
             _config = config;
         }
 
-        private void ListenForClients()
+        private void TryRestart ()
         {
-            ListenerSocket.Accept(OnClientConnect, e => {
-                FleckLog.Error("Listener socket is closed", e);
-                if(RestartAfterListenError){
-                    FleckLog.Info("Listener socket restarting");
-                    try
-                    {
-                        ListenerSocket.Dispose();
-                        var socket = new Socket(_locationIP.AddressFamily, SocketType.Stream, ProtocolType.IP);
-                        ListenerSocket = new SocketWrapper(socket);
-                        Start(_config);
-                        FleckLog.Info("Listener socket restarted");
-                    }
-                    catch (Exception ex)
-                    {
-                        FleckLog.Error("Listener could not be restarted", ex);
-                    }
-                }
-            });
+            FleckLog.Info ("Listener socket restarting");
+            try {
+                ListenerSocket.Dispose ();
+                var socket = new Socket (_locationIP.AddressFamily, SocketType.Stream, ProtocolType.IP);
+                ListenerSocket = new SocketWrapper (socket);
+                Start (_config);
+                FleckLog.Info ("Listener socket restarted");
+            } catch (Exception ex) {
+                FleckLog.Error ("Listener socket could not be restarted", ex);
+            }
+        }
+
+        private void ListenForClients ()
+        {
+          ManualResetEvent acceptDone = new ManualResetEvent (false);
+          bool running = true;
+
+          Task.Run (() => {
+        
+          while (running) {
+          
+              acceptDone.Reset ();
+
+              var task = ListenerSocket.Accept(
+                s => {
+                       running = (s != null);
+                       acceptDone.Set (); 
+                       OnClientConnect (s); },
+                e => { FleckLog.Error ("Error while listening for new clients", e);
+                       if (RestartAfterListenError) TryRestart (); 

```

Comment:
Given this should properly separate listener and client errors, I'm not sure if having the restart is necessary.